dace/autodiff/data_forwarding/store.py

```diff
@@ -0,0 +1,683 @@
+        # Get the size of each loop and add it to the list
+        for loop in all_encolsing_loops:
+            # Get the end of the loop
+            start, end = ad_utils.extract_loop_region_info(loop)
+
+            # Check if the loop is increasing or decreasing
+            # First, try to convert the strings to ints if possible
+            # Note that we look for the start or end of the loop
+            # And not the size of the loop.
+            # This is because we access using the loop indices
+            # Using the loop sizes instead would require shifting accesses
+            _, new_dim = ad_utils.get_loop_end(start, end, loop)
+
+            # First we check if the new dimension contains symbols
+            # These will need to be replaced with scalars for correct allocation
+            # The sdfg symbols are allowed to be in the shape
+            if ad_utils.shape_has_symbols_to_replace(bwd_generator.sdfg, new_dim):
+                # Take the expression to sympy for easier processing
+                if isinstance(new_dim, str):
+                    new_dim = sp.Symbol(new_dim)
+
+                # Try to replace the symbols with the loop size
+                loop_size, loop_index = _get_symbol_upper_bound_from_loop(bwd_generator, new_dim, all_encolsing_loops)
+                if isinstance(loop_size, int) or (isinstance(loop_size, str) and ad_utils.is_int(loop_size)):
+                    new_dim = new_dim.subs(sp.Symbol(loop_index), loop_size)
+                else:
+                    new_dim = new_dim.subs(sp.Symbol(loop_index), dace.symbol(loop_size))
+            shape.insert(0, new_dim)
+            loop_param_list.insert(0, loop.loop_variable)
+
+    # Add the array descriptor and AccessNode to the forward state
+    original_desc = forward_an.desc(forward_state)
+
+    # We make a special case for a memlet of the type A[i, j] in an i, j loop
+    # In this case we only need an array of the same size as the forward node
+    if enclosed and edge.data.data == forward_an.data and len(edge.data.subset) == nb_enclosing_loops:
+        # Check that the memlet subset matches perfectly the order of loop nest
+        # Make sure the subset elements are (i,i,1) and (j,j,1)
+        # Then check if this matches the loop indices
+        if all(
+                str(subset[0]) == loop_param_list[i] and subset[0] == subset[1] and subset[2] == 1
+                for i, subset in enumerate(edge.data.subset)):
+            # We only use the loop accesses
+            # Both should work since shape[:nb_enclosing_loops] == shape[nb_enclosing_loops:]
+            shape = shape[nb_enclosing_loops:]
+
+            # We want to build the memlet as if this was not in a a loop
+            nb_enclosing_loops = 0
+
+    new_store_node = forward_state.add_array(
+        name=new_store_node_name,
+        shape=shape,
+        dtype=original_desc.dtype,
+        transient=True,
+    )
+    new_store_node.setzero = True
```

Comment:
does this mean that if using the store data forwarding, we cannot call the gradients again and again to accumulate data? or are those only intermediate data points that need to be zeroed every time?
This behavior should be documented somewhere outside the code.